### PR TITLE
chore: use `t.Setenv` to set env vars in tests

### PR DESCRIPTION
**Pull request by Juneezee**
Starting from Go 1.17, we can use `t.Setenv` to set environment variable in test. The environment variable is automatically restored to its original value when the test and all its subtests complete. This ensures that each test does not start with leftover environment variables from previous completed tests.

Reference: https://pkg.go.dev/testing#T.Setenv

```go
func TestFoo(t *testing.T) {
	// before
	os.Setenv(key, "new value")
	defer os.Unsetenv(key)
	
	// after
	t.Setenv(key, "new value")
}
```

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: controller/sharding/sharding_test.go [hunk 1]
```diff
@@ -45,22 +45,21 @@ func TestGetShardByID_NoReplicasUsingHashDistributionFunction(t *testing.T) {
...
 	assert.Equal(t, -1, distributionFunction(&cluster2))
 	assert.Equal(t, -1, distributionFunction(&cluster3))
 	assert.Equal(t, -1, distributionFunction(&cluster4))
 	assert.Equal(t, -1, distributionFunction(&cluster5))
-
 }
 
 func TestGetClusterFilterDefault(t *testing.T) {
 	shardIndex := 1 // ensuring that a shard with index 1 will process all the clusters with an "even" id (2,4,6,...)
 	os.Unsetenv(common.EnvControllerShardingAlgorithm)
```
Comment (blakepettersson) [inline controller/sharding/sharding_test.go:61]: Why is this line present? I guess it was due to line 49, but that shouldn't be needed anymore should it?

**Review by blakepettersson — COMMENTED** [review]
One minor nit, otherwise LGTM! 🍾

File: controller/sharding/sharding_test.go (hunk 1 above)
Comment (Juneezee) [reply controller/sharding/sharding_test.go:61]: I think we still need to keep this line. This is to ensure that `ARGOCD_CONTROLLER_SHARDING_ALGORITHM` are not set, which is actually irrelevant of the other tests.

For example, when we do `go test  -run ^TestGetClusterFilterDefault github.com/argoproj/argo-cd/v2/controller/sharding` to run this single test only, this line will ensure that we will never have the env var set.

**Review by blakepettersson — APPROVED** [review]

**Review by crenshaw-dev — APPROVED** [review]
Lovely. Thank you,  @Juneezee!

**Pull request merged**